cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -676,44 +687,55 @@ public void close() throws IOException {
 
   @Override
   public String getInputFormatClassName() {
-    Collection<String> inputPaths = filterInputPaths();
-    if (inputPaths != null && inputPaths.isEmpty()) {
+    Collection<PartitionKey> inputKeys = getInputKeys();
+    if (inputKeys.isEmpty()) {
       return EmptyInputFormat.class.getName();
     }
     return files.getInputFormatClassName();
   }
 
   @Override
   public Map<String, String> getInputFormatConfiguration() {
-    Collection<String> inputPaths = filterInputPaths();
-    if (inputPaths == null) {
-      return files.getInputFormatConfiguration();
+    Collection<PartitionKey> inputKeys = getInputKeys();
+
+    List<Location> inputLocations = new ArrayList<>(inputKeys.size());
+    Map<String, PartitionKey> pathToKey = new HashMap<>(inputKeys.size());
+    for (PartitionKey key : inputKeys) {
+      PartitionDetail partition = getPartition(key);
+      String path = Objects.requireNonNull(partition).getRelativePath();
+      Location partitionLocation = files.getLocation(path);
+      inputLocations.add(partitionLocation);
+      pathToKey.put(partitionLocation.toURI().toString(), key);
     }
-    List<Location> inputLocations = Lists.newArrayListWithExpectedSize(inputPaths.size());
-    for (String path : inputPaths) {
-      inputLocations.add(files.getLocation(path));
-    }
-    return files.getInputFormatConfiguration(inputLocations);
+
+    Map<String, String> inputFormatConfiguration = files.getInputFormatConfiguration(inputLocations);
+    inputFormatConfiguration.put(PATH_TO_PARTITIONING_MAPPING, GSON.toJson(pathToKey));
```

Comment:
this JSON can become really large... say a job consumes thousands of partitions. Is there a limit to the length of a property in the Hadoop conf that this might exceed?